packages/context/src/binding.ts

```diff
@@ -215,17 +215,47 @@ export class Binding<T = BoundValue> {
     return this;
   }
 
-  tag(tagName: string | string[]): this {
-    if (typeof tagName === 'string') {
-      this.tags.add(tagName);
-    } else {
-      tagName.forEach(t => {
-        this.tags.add(t);
-      });
+  /**
+   * Tag the binding with names
+   * @param tagNames A list of names
+   *
+   * @example
+   * ```ts
+   * binding.tag('a');
+   * binding.tag('a', 'b');
+   * binding.tag(['a', 'b'], 'c')
+   * ```
+   */
+  tag(...tagNames: (string | string[])[]): this {
+    for (const tagName of tagNames) {
+      if (typeof tagName === 'string') {
+        this.tagValue(tagName, true);
+      } else {
+        tagName.forEach(t => {
+          this.tagValue(t, true);
+        });
+      }
     }
     return this;
   }
 
+  /**
+   * Tag the binding with name and value
+   * @param name Name of the tag
+   * @param value Value of the tag
+   */
+  tagValue(name: string, value: BoundValue): this {
+    this.tagValues.set(name, value);
+    return this;
+  }
+
+  /**
+   * Get a set of tag names
+   */
+  get tags() {
+    return new Set<string>(this.tagValues.keys());
```

Comment:
Is there any particular reason why you are converting the result of `tagValues.keys()` to a `Set`? Cannot we simply return whatever `keys` returned and let the caller decide how they want to work with that? Allocating a new Set whenever we are querying binding tags may hurt performance.

Reply:
It was `Set` before to avoid duplicate entries. Now we can use `string[]` as the type.